flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -1195,53 +1189,72 @@ private ApplicationReport startAppMaster(
         yarnClient.submitApplication(appContext);
 
         LOG.info("Waiting for the cluster to be allocated");
+
+        try {
+            waitTillTargetState(
+                    yarnClient, appId, YarnApplicationState.ACCEPTED, YarnApplicationState.RUNNING);
+        } catch (IOException e) {
+            throw new YarnDeploymentException("Failed to deploy the cluster.", e);
+        }
+
+        // since deployment was successful, remove the hook
+        ShutdownHookUtil.removeShutdownHook(deploymentFailureHook, getClass().getSimpleName(), LOG);
+
+        return ApplicationReportProviderImpl.of(yarnClient, applicationId);
```

Comment:
Aren't we leaking the `yarnClient` here? How will the lifecycle of the `yarnClient` be managed?

Reply:
Maybe the `yarnClient` should be used directly in the method of `waitTillTargetState`, not as the param of method.
Right?